### PR TITLE
Implement unsubscribe

### PR DESCRIPTION
**Pull request by woss**

### CODE REVIEW

File: core/client.go [hunk 1]
```diff
@@ -104,7 +104,12 @@ func (c *client) Publish(message *messages.Publish, data []byte) {
 	if debug {
 		log.Printf("Client from %s publishing %s to %s", c.GetRemoteAddr(), message.Content, message.Channel)
 	}
-
+	// Don't publish to a channel that client is not subscribed
```
Comment (lietu) [inline core/client.go]: Remove, there is no need for a client to subscribe to a channel before publishing to it.

File: core/client.go [hunk 2]
```diff
@@ -152,6 +157,22 @@ func (c *client) IsSubscribed(channel string) bool {
 	return false
 }
 
+func (c *client) Unsubscribe(message *messages.Unsubscribe) {
+	if debug {
+		log.Printf("Client from %s unsubscribing from %s", c.GetRemoteAddr(), message.Channel)
+	}
+
+	// Return
+	if !c.IsSubscribed(message.Channel) {
```
Comment (lietu) [inline core/client.go:150]: No need to error here.

**Comment by woss** [comment]
remove the unsubscribe_ok response, just throw clients out

File: core/client.go [hunk 3]
```diff
@@ -152,6 +157,22 @@ func (c *client) IsSubscribed(channel string) bool {
...
+		c.SendMessage(messages.NewGenericMessage(messages.TYPE_UNSUBSCRIBABLE_CHANNEL))
+		c.Close()
+		return
+	}
+
+	unsubscribe(message.Channel, c)
+	c.SendMessage(messages.NewGenericMessage(messages.TYPE_UNSUBSCRIBE_OK))
```
Comment (lietu) [inline core/client.go]: We don't need the `unsubscribe_ok` event, there is no need to make the clients wait for that.

File: core/client.go (hunk 1 above)
Comment (woss) [inline core/client.go]: done

File: core/client.go (hunk 2 above)
Comment (woss) [inline core/client.go:150]: removed

File: core/client.go (hunk 3 above)
Comment (woss) [inline core/client.go]: removed

File: core/client.go [hunk 4]
```diff
@@ -10,7 +10,7 @@ import (
 	"net/http"
 )
 
-const debug = false
+const debug = true
```
Comment (lietu) [inline core/client.go]: please don't commit this

File: core/client.go [hunk 5]
```diff
@@ -152,6 +152,13 @@ func (c *client) IsSubscribed(channel string) bool {
 	return false
 }
 
+func (c *client) Unsubscribe(message *messages.Unsubscribe) {
+	if debug {
+		log.Printf("Client from %s unsubscribing from %s", c.GetRemoteAddr(), message.Channel)
+	}
+	unsubscribe(message.Channel, c)
```
Comment (lietu) [inline core/client.go]: Now this doesn't remove the channel from `c.Subscriptions`?

File: messages/clientmessages.go [hunk 6]
```diff
@@ -56,6 +56,11 @@ type Subscribe struct {
 	Channel string `json:"channel"`
 }
 
+type Unsubscribe struct {
```
Comment (lietu) [inline messages/clientmessages.go]: It would be nice if the struct + it's methods + new function would be grouped next to each other, not scattered around the file

File: core/client.go (hunk 4 above)
Comment (woss) [inline core/client.go]: yup, my bad

File: messages/clientmessages.go (hunk 6 above)
Comment (woss) [inline messages/clientmessages.go]: done

File: core/subscriptions.go [hunk 7]
```diff
@@ -26,8 +26,6 @@ func subscribe(channel string, client *client) {
 func unsubscribe(channel string, client *client) {
 	filtered := clients{}
 
-	// Probably can't handle unsubscribing from a non-existent channel, but
```
Comment (lietu) [inline core/subscriptions.go:29]: Why did you remove these?

Comment (woss) [inline core/subscriptions.go:29]: because it is usless to check that. Unsubscribe method checks if person is unsubscribing from not subscribed channel, never ends up there

File: core/client.go [hunk 8]
```diff
@@ -152,6 +152,28 @@ func (c *client) IsSubscribed(channel string) bool {
 	return false
 }
 
+func (c *client) Unsubscribe(message *messages.Unsubscribe) {
+	if debug {
+		log.Printf("Client from %s unsubscribing from %s", c.GetRemoteAddr(), message.Channel)
+	}
+
+	if !c.IsSubscribed(message.Channel) {
+		return
```
Comment (woss) [inline core/client.go:151]: Here is checking is it subscribed or not

File: core/subscriptions.go (hunk 7 above)
Comment (lietu) [inline core/subscriptions.go:29]: IT's not checking anything, it's a comment stating something that hasn't changed since the comment was written.

Comment (woss) [inline core/subscriptions.go:29]: alright, i removed it because if a person is not subscribed triggering unsubscribe should never end up there. if you wish i will put it back.
